[BUG] StreamableHttpMcpTransport doesn't renew a lost Mcp-Session-Id

**Describe the bug**

LangChain4j now has the `StreamableHttpMcpTransport` transport for streamable-http MCP protocol (https://github.com/langchain4j/langchain4j/pull/3534). 

It supports the `Mcp-Session-Id` header that tracks the MCP session, but it doesn't handle the case where the server loses/terminates the session.

**Log and Stack trace**

```
java.util.concurrent.ExecutionException: java.lang.RuntimeException: Unexpected status code: 404
        at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396) ~[na:na]
        at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2096) ~[na:na]
        at dev.langchain4j.mcp.client.DefaultMcpClient.checkHealth(DefaultMcpClient.java:329) ~[langchain4j-mcp-1.6.0-beta12.jar:na]
        ... 82 common frames omitted
Caused by: java.lang.RuntimeException: Unexpected status code: 404
        at dev.langchain4j.mcp.client.transport.http.StreamableHttpMcpTransport$1.apply(StreamableHttpMcpTransport.java:127) ~[langchain4j-mcp-1.6.0-beta12.jar:na]
        at java.net.http/jdk.internal.net.http.Http1Exchange.createResponseSubscriber(Http1Exchange.java:460) ~[java.net.http:na]
        at java.net.http/jdk.internal.net.http.Http1Exchange.readBodyAsync(Http1Exchange.java:451) ~[java.net.http:na]
        at java.net.http/jdk.internal.net.http.Exchange.readBodyAsync(Exchange.java:225) ~[java.net.http:na]
        at java.net.http/jdk.internal.net.http.MultiExchange.lambda$responseAsync0$4(MultiExchange.java:367) ~[java.net.http:na]
        at java.base/java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1150) ~[na:na]
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510) ~[na:na]
        at java.base/java.util.concurrent.CompletableFuture.postFire(CompletableFuture.java:614) ~[na:na]
        at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:653) ~[na:na]
        at java.base/java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:482) ~[na:na]
        at java.net.http/jdk.internal.net.http.HttpClientImpl$DelegatingExecutor.execute(HttpClientImpl.java:177) ~[java.net.http:na]
        at java.base/java.util.concurrent.CompletableFuture$UniCompletion.claim(CompletableFuture.java:572) ~[na:na]
        at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642) ~[na:na]
        at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510) ~[na:na]
        at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179) ~[na:na]
        at java.net.http/jdk.internal.net.http.Http1Response$HeadersReader.handle(Http1Response.java:610) ~[java.net.http:na]
        at java.net.http/jdk.internal.net.http.Http1Response$HeadersReader.handle(Http1Response.java:536) ~[java.net.http:na]
        at java.net.http/jdk.internal.net.http.Http1Response$Receiver.accept(Http1Response.java:527) ~[java.net.http:na]
        at java.net.http/jdk.internal.net.http.Http1Response$HeadersReader.tryAsyncReceive(Http1Response.java:583) ~[java.net.http:na]
        at java.net.http/jdk.internal.net.http.Http1AsyncReceiver.flush(Http1AsyncReceiver.java:233) ~[java.net.http:na]
        at java.net.http/jdk.internal.net.http.common.SequentialScheduler$LockingRestartableTask.run(SequentialScheduler.java:182) ~[java.net.http:na]
        at java.net.http/jdk.internal.net.http.common.SequentialScheduler$CompleteRestartableTask.run(SequentialScheduler.java:149) ~[java.net.http:na]
        at java.net.http/jdk.internal.net.http.common.SequentialScheduler$SchedulableTask.run(SequentialScheduler.java:207) ~[java.net.http:na]
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[na:na]
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[na:na]
        at java.base/java.lang.Thread.run(Thread.java:1583) ~[na:na]
```

**To Reproduce**

1. Initialize a `DefaultMcpClient` with a `StreamableHttpMcpTransport`.
2. Use it to connect to an MCP server that handles MCP sessions (e.g. one based on Python-SDK: https://pypi.org/project/mcp/). 
3. A simple health check should suffice to spawn a new session. LangChain4j will now start sending `Mcp-Session-Id` in the headers.
4. Make the server lose the session (e.g. by restarting the mentioned server).
5. Try to contact the MCP server via the client again, e.g. run another health check.

**Current behavior**
A `RuntimeException` is thrown. It's caused by HTTP reply with an error code 4xx.

`StreamableHttpMcpTransport.java`:
```java
if (!isExpectedStatusCode(responseInfo.statusCode())) {
    future.completeExceptionally(
            new RuntimeException("Unexpected status code: " + responseInfo.statusCode()));
    return null;
}
```
```java
private boolean isExpectedStatusCode(int statusCode) {
    return statusCode >= 200 && statusCode < 300;
}
```

**Expected behavior**
Some type of handling for this case should be provided. The spec's point 2.5. [1] says:

> 3. The server **MAY** terminate the session at any time, after which it **MUST** respond to requests containing that session ID with HTTP 404 Not Found.
> 4. When a client receives HTTP 404 in response to a request containing an `Mcp-Session-Id`, it **MUST** start a new session by sending a new `InitializeRequest` without a session ID attached.

Whether this should happen automatically or with some intervention from the library's user -- I'm not sure.

Also, the lib should not just throw `RuntimeException` under any circumstance, please very much.

**Please complete the following information:**
- LangChain4j version: 1.6.0

**Additional context:**
There's currently a separate bug in the Python SDK where it returns "400 Bad Request" whereas the spec says "404 Not Found" should be returned (https://github.com/modelcontextprotocol/python-sdk/issues/1004). I pushed a PR to fix that (https://github.com/modelcontextprotocol/python-sdk/pull/1417).

**Reference:**
1. https://modelcontextprotocol.io/specification/2025-06-18/basic/transports#session-management

--- a/langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/StreamableHttpMcpTransport.java
+++ b/langchain4j-mcp/src/main/java/dev/langchain4j/mcp/client/transport/http/StreamableHttpMcpTransport.java
@@ -37,9 +37,10 @@ public class StreamableHttpMcpTransport implements McpTransport {
     private final boolean logRequests;
     private final Logger trafficLog;
     static final ObjectMapper OBJECT_MAPPER = new ObjectMapper();
-
+    private final AtomicReference<CompletableFuture<JsonNode>> initializeInProgress = new AtomicReference<>(null);
     private volatile McpOperationHandler operationHandler;
     private final HttpClient httpClient;
+    private McpInitializeRequest initializeRequest;
     private final AtomicReference<String> mcpSessionId = new AtomicReference<>();
 
     public StreamableHttpMcpTransport(StreamableHttpMcpTransport.Builder builder) {
@@ -63,7 +64,14 @@ public class StreamableHttpMcpTransport implements McpTransport {
 
     @Override
     public CompletableFuture<JsonNode> initialize(McpInitializeRequest operation) {
-        return execute(operation, operation.getId())
+        this.initializeRequest = operation;
+        CompletableFuture<JsonNode> completableFuture = execute(operation, operation.getId());
+        initializeInProgress.set(completableFuture);
+        return completableFuture
+                .thenCompose(originalResponse -> {
+                    initializeInProgress.set(null);
+                    return CompletableFuture.completedFuture(originalResponse);
+                })
                 .thenCompose(originalResponse -> execute(new McpInitializationNotification(), null)
                         .thenCompose(nullNode -> CompletableFuture.completedFuture(originalResponse)));
     }
@@ -76,7 +84,7 @@ public class StreamableHttpMcpTransport implements McpTransport {
         }
         final HttpRequest.Builder builder = HttpRequest.newBuilder();
         String sessionId = mcpSessionId.get();
-        if (sessionId != null) {
+        if (sessionId != null && !(message instanceof McpInitializeRequest)) {
             builder.header("Mcp-Session-Id", sessionId);
         }
         customHeaders.forEach(builder::header);
@@ -108,6 +116,14 @@ public class StreamableHttpMcpTransport implements McpTransport {
     }
 
     private CompletableFuture<JsonNode> execute(McpClientMessage message, Long id) {
+        return execute(message, id, false);
+    }
+
+    private CompletableFuture<JsonNode> execute(McpClientMessage message, Long id, boolean isRetry) {
+        CompletableFuture<JsonNode> reinitializeInProgress = this.initializeInProgress.get();
+        if (reinitializeInProgress != null) {
+            reinitializeInProgress.join();
+        }
         HttpRequest request = null;
         try {
             request = createRequest(message);
@@ -122,9 +138,27 @@ public class StreamableHttpMcpTransport implements McpTransport {
         httpClient
                 .sendAsync(request, responseInfo -> {
                     if (!isExpectedStatusCode(responseInfo.statusCode())) {
-                        future.completeExceptionally(
-                                new RuntimeException("Unexpected status code: " + responseInfo.statusCode()));
-                        return null;
+                        if (!(message instanceof McpInitializeRequest) && responseInfo.statusCode() == 404) {
+                            if (!isRetry) {
+                                initialize(StreamableHttpMcpTransport.this.initializeRequest)
+                                        .thenAccept(node -> {
+                                            execute(message, id, true)
+                                                    .thenAccept(future::complete)
+                                                    .exceptionally(t -> {
+                                                        future.completeExceptionally(t);
+                                                        return null;
+                                                    });
+                                        })
+                                        .exceptionally(t -> {
+                                            future.completeExceptionally(t);
+                                            return null;
+                                        });
+                            }
+                        } else {
+                            future.completeExceptionally(
+                                    new RuntimeException("Unexpected status code: " + responseInfo.statusCode()));
+                        }
+                        return HttpResponse.BodySubscribers.discarding();
                     } else {
                         Optional<String> contentType = responseInfo.headers().firstValue("Content-Type");
                         Optional<String> mcpSessionId = responseInfo.headers().firstValue("Mcp-Session-Id");

--- a/langchain4j-mcp/src/test/java/dev/langchain4j/mcp/client/integration/McpServerHelper.java
+++ b/langchain4j-mcp/src/test/java/dev/langchain4j/mcp/client/integration/McpServerHelper.java
@@ -6,16 +6,8 @@ import static org.junit.jupiter.api.Assumptions.assumeTrue;
 import java.io.File;
 import java.io.IOException;
 import java.util.Arrays;
-import java.util.List;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.TimeoutException;
-
-import dev.langchain4j.agent.tool.ToolSpecification;
-import dev.langchain4j.mcp.McpToolProvider;
-import dev.langchain4j.mcp.client.McpClient;
-import dev.langchain4j.service.tool.ToolExecutor;
-import dev.langchain4j.service.tool.ToolProvider;
-import dev.langchain4j.service.tool.ToolProviderResult;
 import okhttp3.OkHttpClient;
 import okhttp3.Request;
 import org.slf4j.Logger;
@@ -29,10 +21,12 @@ public class McpServerHelper {
         return startServerHttp(scriptName, 8080);
     }
 
-    static Process startServerHttp(String scriptName, int port) throws InterruptedException, TimeoutException, IOException {
+    static Process startServerHttp(String scriptName, int port)
+            throws InterruptedException, TimeoutException, IOException {
         skipTestsIfJbangNotAvailable();
         String path = getPathToScript(scriptName);
-        String[] command = new String[] {getJBangCommand(), "--quiet", "--fresh", "run", "-Dquarkus.http.port=" + port, path};
+        String[] command =
+                new String[] {getJBangCommand(), "--quiet", "--fresh", "run", "-Dquarkus.http.port=" + port, path};
         log.info("Starting the MCP server using command: " + Arrays.toString(command));
         Process process = new ProcessBuilder().command(command).inheritIO().start();
         waitForPort(port, 120);

--- a/langchain4j-mcp/src/test/java/dev/langchain4j/mcp/client/integration/McpToolsStreamableHttpTransportIT.java
+++ b/langchain4j-mcp/src/test/java/dev/langchain4j/mcp/client/integration/McpToolsStreamableHttpTransportIT.java
@@ -3,14 +3,18 @@ package dev.langchain4j.mcp.client.integration;
 import static dev.langchain4j.mcp.client.integration.McpServerHelper.skipTestsIfJbangNotAvailable;
 import static dev.langchain4j.mcp.client.integration.McpServerHelper.startServerHttp;
 
+import dev.langchain4j.agent.tool.ToolSpecification;
 import dev.langchain4j.mcp.client.DefaultMcpClient;
 import dev.langchain4j.mcp.client.transport.McpTransport;
 import dev.langchain4j.mcp.client.transport.http.StreamableHttpMcpTransport;
 import java.io.IOException;
 import java.time.Duration;
+import java.util.List;
 import java.util.concurrent.TimeoutException;
 import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -42,5 +46,27 @@ class McpToolsStreamableHttpTransportIT extends McpToolsTestBase {
         if (process != null && process.isAlive()) {
             process.destroyForcibly();
         }
+    }
+
+    @Test
+    void reinitializesSessionWhenMissingFromServer() throws Exception {
+        // Initial call to establish session and verify MCP server is up
+        List<ToolSpecification> toolSpecifications = mcpClient.listTools();
+        Assertions.assertNotNull(toolSpecifications, "Tool specifications should not be null");
+        Assertions.assertDoesNotThrow(() -> mcpClient.checkHealth(), "Health check should pass initially");
+
+        // Simulate server crash
+        process.destroyForcibly();
+
+        // Expect failure as server is down
+        Assertions.assertThrows(
+                Exception.class, () -> mcpClient.checkHealth(), "Expected exception when server is down");
+
+        // Restart the server
+        process = startServerHttp("tools_mcp_server.java");
+
+        // Subsequent call should reinitialize session and succeed
+        Assertions.assertDoesNotThrow(
+                () -> mcpClient.checkHealth(), "Session should be reinitialized and health check should pass");
     }
 }

